src/evm/token.rs

```diff
@@ -0,0 +1,248 @@
+use crate::protocol::vm::{
+    constants::EXTERNAL_ACCOUNT, erc20_overwrite_factory::ERC20OverwriteFactory,
+    tycho_simulation_contract::TychoSimulationContract, utils::ERC20Slots,
+};
+use ethers::{
+    abi::{Abi, Token},
+    types::{H160, U256},
+};
+use lazy_static::lazy_static;
+use revm::{primitives::Address, DatabaseRef};
+use serde_json::from_str;
+use std::str::FromStr;
+use thiserror::Error;
+
+use super::{simulation::SimulationEngine, simulation_db::BlockHeader, ContractCompiler};
+
+const MARKER_VALUE: u128 = 3141592653589793238462643383;
+const SPENDER: &str = "08d967bb0134F2d07f7cfb6E246680c53927DD30";
+lazy_static! {
+    static ref ERC20_ABI: Abi = {
+        let abi_file_path = "src/protocol/vm/assets/ERC20.abi";
+        let abi_json = std::fs::read_to_string(abi_file_path).expect("Failed to read ABI file");
+        from_str(&abi_json).expect("Failed to parse ABI JSON")
+    };
+}
+
+/// An error r
+#[derive(Error, Debug)]
+pub enum TokenError {
```

Comment:
Is it overkill? I can make it a `String` but I like to have specific types for errors. Or if you have any suggestion about where this error could be added?

Reply:
I think this has to be a `SimulationError`